src/Language/Fixpoint/Solver/Interpreter.hs

```diff
@@ -0,0 +1,856 @@
+  where
+    go []  []     = PTrue
+    go [p] []     = interpret' ie γ ctx env p
+    go acc []     = PAnd acc
+    go acc (e:es) = if e == PTrue then go acc es
+                                  else if e == PFalse then PFalse else go (e:acc) es
+interpret ie γ ctx env (POr es)         = let es' = map (interpret' ie γ ctx env) es in go [] (reverse es')
+  where
+    go []  []     = PFalse
+    go [p] []     = interpret' ie γ ctx env p
+    go acc []     = POr acc
+    go acc (e:es) = if e == PTrue then PTrue
+                                  else if e == PFalse then go acc es else go (e:acc) es
+interpret ie γ ctx env (PNot e)         = let e' = interpret' ie γ ctx env e in case e' of
+    (PNot e'')    -> e''
+    PTrue         -> PFalse 
+    PFalse        -> PTrue 
+    _             -> PNot e'
+interpret ie γ ctx env (PImp e1 e2)     = let e1' = interpret' ie γ ctx env e1 
+                                              e2' = interpret' ie γ ctx env e2 in
+                                            if e1' == PFalse || e2' == PTrue then PTrue else
+                                              if e1' == PTrue then e2' else
+                                                if e2' == PFalse then interpret' ie γ ctx env (PNot e1') else 
+                                                  PImp e1' e2'
+interpret ie γ ctx env (PIff e1 e2)     = let e1' = interpret' ie γ ctx env e1 
+                                              e2' = interpret' ie γ ctx env e2 in
+                                            if e1' == PTrue then e2' else
+                                              if e2' == PTrue then e1' else
+                                                if e1' == PFalse then interpret' ie γ ctx env (PNot e2') else
+                                                  if e2' == PFalse then interpret' ie γ ctx env (PNot e1') else
+                                                    PIff e1' e2'
+interpret ie γ ctx env (PAtom o e1 e2)  = let e1' = interpret' ie γ ctx env e1
+                                              e2' = interpret' ie γ ctx env e2 in
+                                            applyBooleanFolding o e1' e2'
+interpret _  _ _   _   e@(PKVar _ _)    = e
+interpret ie γ ctx env e@(PAll xss e1)  = case xss of
+  [] -> interpret' ie γ ctx env e1
+  _  -> e
+interpret ie γ ctx env e@(PExist xss e1) = case xss of
+  [] -> interpret' ie γ ctx env e1
+  _  -> e
+interpret _  _ _   _   e@(PGrad _ _ _ _) = e
+interpret ie γ ctx env (ECoerc s t e)    = let e' = interpret' ie γ ctx env e in
+                                             if s == t then e' else (ECoerc s t e')
+
+        
+--------------------------------------------------------------------------------
+-- | Knowledge (SMT Interaction)
+--------------------------------------------------------------------------------
+data Knowledge = KN 
+  { knSims              :: M.HashMap (Symbol, Symbol) Rewrite  -- ^ Rewrite rules came from match and data type definitions 
```

Comment:
`knSims` was changed in the smt-based PLE to `Map Symbol Rewrite` which allows faster lookups there.

At this point I'm not sure if it is worth insisting on porting every change. Maybe we should just agree to address the differences whenever we factorize the common parts of both PLEs.